language/move-vm/runtime/src/runtime.rs

```diff
@@ -73,14 +77,35 @@ impl VMRuntime {
             .finish(Location::Undefined));
         }
 
-        // Make sure that there is not already a module with this name published
-        // under the transaction sender's account.
         let module_id = compiled_module.self_id();
-        if data_store.exists_module(&module_id)? {
-            return Err(
-                PartialVMError::new(StatusCode::DUPLICATE_MODULE_NAME).finish(Location::Undefined)
-            );
-        };
+
+        if !allow_republish {
+            // Make sure that there is not already a module with this name published under the
+            // transaction sender's account.
+            if data_store.exists_module(&module_id)? {
+                return Err(PartialVMError::new(StatusCode::DUPLICATE_MODULE_NAME)
+                    .finish(Location::Undefined));
+            };
+        } else if !allow_breaking_changes {
+            // Make sure that the to-be-published module do not introduce breaking changes if there
+            // exists an old module with the same module id in data store
+            if let Ok(old_module_bytes) = data_store.load_module(&module_id) {
+                let old_module = match CompiledModule::deserialize(&old_module_bytes) {
+                    Ok(module) => module,
+                    Err(err) => {
+                        warn!(*log_context, "[VM] module deserialization failed {:?}", err);
+                        return Err(err.finish(Location::Undefined));
+                    }
+                };
+                let old_m = Module::new(&old_module);
```

Comment:
Nit: writing `normalized::Module` might make the difference between this type and `CompiledModule` a bit more explicit.